### DISCUSSION TITLE
Changes to ZHA, zigpy, and zhaquirks to allow custom quirks

### DISCUSSION THREAD

**Discussion by puddly** [zigpy/zigpy#603] [RFC]
Modifying the installed `zhaquirks` module directly inside of `site-packages` is a little unwieldy and doesn't survive package upgrades, making it quite difficult for new users to work with. I think it'd be useful to have something like this be possible:

```yaml
zha:
  zigpy:
    custom_quirks: /config/custom_zha_quirks/  # or something similar, it doesn't really matter
```

The `custom_zha_quirks` folder can be like an external `zhaquirks` folder whose submodules take priority over ones contained within the `zhaquirks` in `site-packages` and survive Home Assistant upgrades.

Right now the implicit, metaclass-based quirk registration (i.e. subclassing creates the quirk) is useful but also has the downside of user-created custom quirks not being able to import clusters from existing quirk submodules without also registering them, making it impossible to "subclass" existing quirks to tweak a few things. For example:

```Python
import copy

from zhaquirks.const import ENDPOINTS, INPUT_CLUSTERS
from zhaquirks.xiaomi import XiaomiPowerConfiguration
from zhaquirks.xiaomi.aqara.motion_aq2 import MotionAQ2

XIAOMI_BATTERY_LUT = {...}

class BetterXiaomiPowerConfiguration(XiaomiPowerConfiguration):
    def _calculate_battery_percentage(self, raw_value):
        return XIAOMI_BATTERY_LUT[raw_value]
    

class CustomPowerMotionAQ2(MotionAQ2):
    replacement = copy.deepcopy(MotionAQ2.replacement)
    replacement[ENDPOINTS][1][INPUT_CLUSTERS].replace(
        XiaomiPowerConfiguration,
        BetterXiaomiPowerConfiguration
    )
```

`zhaquirks` currently relies on import-time quirk registration and is never actually imported by `zigpy` itself, which makes the `zha:zigpy:custom_quirks` configuration option impossible to implement without changing all three packages.

How do the following changes sound?

 1. Replace metaclass quirk registration with a single function: `zigpy.quirks.register(custom_device)`.
 2. Create a module-level `zhaquirks.setup(zigpy_config)` function that can be called by ZHA with the `zha:zigpy:` configuration dictionary.
 3. `zhaquirks.setup` would then look for all quirks (i.e. any exported `CustomDevice` subclasses) and register them in the appropriate order via `zigpy.quirks.register`, with custom quirks coming first, and then the normal quirks.

**Comment by Adminiuga** [comment]
Just load your custom quirks last from ZHA and it should work?
i.e. just do what zha-quirks does in https://github.com/zigpy/zha-device-handlers/blob/6c480990876cbbe1590efe2cfe5e49fd7cd07a02/zhaquirks/__init__.py#L383-L386 in ZHA __init__.py?

**Reply by puddly** [reply]
Yep, this is what I'm doing at the moment. Just gotta get to finishing up the unit tests.
